pkg/controllers/subnetset/subnetset_controller.go

```diff
@@ -71,8 +74,44 @@ func (r *SubnetSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		log.Error(err, "Unable to fetch SubnetSet CR", "SubnetSet", req.NamespacedName)
 		return ResultRequeue, err
 	}
+
+	bindingCRs := r.getSubnetBindingCRsBySubnetSet(ctx, subnetsetCR)
+	if len(bindingCRs) > 0 {
+		if !controllerutil.ContainsFinalizer(subnetsetCR, servicecommon.SubnetSetFinalizerName) {
+			controllerutil.AddFinalizer(subnetsetCR, servicecommon.SubnetSetFinalizerName)
+			if err := r.Client.Update(ctx, subnetsetCR); err != nil {
+				log.Error(err, "Failed to add the finalizer", "SubnetSet", req.NamespacedName)
+				msgFailAddFinalizer := fmt.Sprintf("Failed to add the finalizer on SubnetSet for the dependency by SubnetConnectionBindingMap %s", bindingCRs[0].Name)
+				r.StatusUpdater.UpdateFail(ctx, subnetsetCR, err, "Unable to add the finalizer on SubnetSet used by SubnetConnectionBindingMap",
+					setSubnetSetReadyStatusFalse, msgFailAddFinalizer)
+				return ResultRequeue, err
+			}
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(subnetsetCR, servicecommon.SubnetSetFinalizerName) {
+			controllerutil.RemoveFinalizer(subnetsetCR, servicecommon.SubnetSetFinalizerName)
+			if err := r.Client.Update(ctx, subnetsetCR); err != nil {
+				log.Error(err, "Failed to delete the finalizer", "SubnetSet", req.NamespacedName)
+				msgFailDelFinalizer := "Failed to remove the finalizer on SubnetSet when there is no reference by SubnetConnectionBindingMaps"
+				r.StatusUpdater.UpdateFail(ctx, subnetsetCR, err, "Unable to remove the finalizer from SubnetSet",
+					setSubnetSetReadyStatusFalse, fmt.Sprint(msgFailDelFinalizer))
+				return ResultRequeue, err
+			}
+		}
+	}
+
 	if !subnetsetCR.ObjectMeta.DeletionTimestamp.IsZero() {
 		r.StatusUpdater.IncreaseDeleteTotal()
+		bindingsOnNSX := r.getNSXSubnetBindingsBySubnetSet(string(subnetsetCR.UID))
+		if len(bindingsOnNSX) > 0 {
+			err := fmt.Errorf("SubnetSet %s is used by SubnetConnectionBindingMaps %s", req.String(), bindingsOnNSX[0].GetName())
+			log.Error(err, "Failed to delete SubnetSet CR because it is used by SubnetConnectionBindingMaps, retrying", "SubnetSet", req.NamespacedName)
+			r.StatusUpdater.DeleteFail(req.NamespacedName, nil, err)
+			msgDeleteInUse := fmt.Sprintf("SubnetSet is used by SubnetConnectionBindingMap %s and not able to delete", bindingsOnNSX[0].GetName())
```

Comment:
if multiple bindings found for the subnetSet, should we show all the bindings instead of just the first one?

Reply:
I don't think it is necessary to show all.